File: src/structure/createTargetFile.ts
import {Configuration} from "../configuration/configuration.js";
import {generateImports} from "./generateImports.js";
import {StructureItem} from "./structure.js";
import {packageToOutputFileName} from "./package/packageToFileName.js";
import {createPackageName} from "./package/createPackageName.js";
import {removeUnusedImports} from "./import/removeUnusedImports.js";

export function createTargetFile(
    item: StructureItem,
    body: string,
    configuration: Configuration,
) {
    const {
        fileName,
        package: packageChunks,
        moduleName,
        qualifier,
        hasRuntime,
        imports,
    } = item

    const {granularity, disclaimer} = configuration

    const packageName = createPackageName(packageChunks)

    const outputFileName = packageToOutputFileName(packageChunks, fileName, configuration)

    const resultImports = removeUnusedImports(imports, body)
        .concat(generateImports(outputFileName, configuration))
        .filter(Boolean)
        .join("\n")

    const jsModule = hasRuntime ? `@file:JsModule("${moduleName}")` : ""
    const jsQualifier = hasRuntime && qualifier ? `@file:JsQualifier("${qualifier}")` : ""
    const typeAliasSuppress = granularity === "top-level" ? "" : `
@file:Suppress(
    "NON_EXTERNAL_DECLARATION_IN_INAPPROPRIATE_FILE",
)
    `.trim()

    const fileAnnotations = [
        jsModule,
        jsQualifier,
        typeAliasSuppress
    ]
        .filter(Boolean)
        .join("\n")

    const content = [
        disclaimer,
        fileAnnotations,
        `package ${packageName}`,
        resultImports,
        body,
    ]
        .filter(Boolean)
        .join("\n\n")

    return content.trim() + "\n"
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/structure/createTargetFile.kt
package io.github.sgrishchenko.karakum.structure

import io.github.sgrishchenko.karakum.configuration.Configuration
import io.github.sgrishchenko.karakum.configuration.Granularity
import io.github.sgrishchenko.karakum.structure.import.removeUnusedImports
import io.github.sgrishchenko.karakum.structure.`package`.createPackageName
import io.github.sgrishchenko.karakum.structure.`package`.packageToOutputFileName

fun createTargetFile(
    item: StructureItem,
    body: String,
    configuration: Configuration,
): String {
    val fileName = item.fileName
    val packageChunks = item.`package`
    val moduleName = item.moduleName
    val qualifier = item.qualifier
    val hasRuntime = item.hasRuntime
    val imports = item.imports

    val granularity  = configuration.granularity
    val disclaimer = configuration.disclaimer

    val packageName = createPackageName(packageChunks)

    val outputFileName = packageToOutputFileName(packageChunks, fileName, configuration)

    val resultImports = (
            removeUnusedImports(imports, body) +
                    generateImports(outputFileName, configuration)
            )
        .filter { it.isNotEmpty() }
        .joinToString(separator = "\n")

    val jsModule = if (hasRuntime) "@file:JsModule(\"${moduleName}\")" else ""
    val jsQualifier = if (hasRuntime && qualifier != null) "@file:JsQualifier(\"${qualifier}\")" else ""
    val typeAliasSuppress = if (granularity == Granularity.topLevel) "" else """
@file:Suppress(
    "NON_EXTERNAL_DECLARATION_IN_INAPPROPRIATE_FILE",
)
    """.trim()

    val fileAnnotations = arrayOf(
        jsModule,
        jsQualifier,
        typeAliasSuppress
    )
        .filter { it.isNotEmpty() }
        .joinToString(separator = "\n")

    val content = arrayOf(
        disclaimer,
        fileAnnotations,
        "package $packageName",
        resultImports,
        body,
    )
        .filter { it.isNotEmpty() }
        .joinToString(separator = "\n\n")

    return content.trim() + "\n"
}
